packages/aws_signature_v4/lib/src/request/canonical_request/canonical_request_util.dart

```diff
@@ -3,11 +3,16 @@
 
 part of 'canonical_request.dart';
 
+final RegExp _encodedComponent = RegExp('%[A-F0-9]{2}');
+
 /// Encodes a query parameter while preventing double-encoding.
 String _safeEncode(String queryComponent) {
```

Comment:
Not fully in scope of these changes, but I wonder if we are handling all edge cases appropriately with this. Two edge cases come to mind:
1. Strings that have a %XX pattern in them, but are not actually encoded. For example `foo bar%20`. This function will leave the string unchanged. Should this be encoded to `foo%20bar%2520`?
2. Strings that are double/triple/etc. encoded. For example,`foo%2520bar` (which is `foo bar` double encoded). Should these be fully decoded, and then single encoded? Or should we leave them encoded multiple times?